### DISCUSSION TITLE
Oracle ODBC: get(T) caused exception "bad cast"

### DISCUSSION THREAD

**Discussion by 330305020** [SOCI/soci#1077] [Q&A] [answered]
I'm using oracle ODBC. The exception "bad cast" is happened in "get" function in "from_base", "p.id= v.get<int>("ID")". SOCI software think it should be type "double" from database, but actually I created table with field "id" as integer, not double. How to solve this problem? I need your help.

My case is below:
struct person
{
int id;
int class;
char name[10];
};

// create table
sql << create table AA (id integer not null, class integer, name varchar2(10), constraint pk_AA primary key (id));

namespace soci
{
  // name-based conversion
  template<> struct type_conversion<person>
  {
    typedef values base_type;

    static void from_base(values const &v, indicator /* ind */, person &p)
    {
      // ignoring possibility that the whole object might be NULL

      p.id= v.get<int>("ID");
      p.class= v.get<int>("CLASS", 0);
      strcpy(p.name, v.get<std::string>("NAME").c_str());
    }

    static void to_base(person const & p, values & v, indicator & ind)
    {
      v.set("ID", p.id);
      v.set("CLASS", p.class);

      std::string str1 = p.name;
      v.set("NAME", str1);
    }
  };
}

person p1;
sql << "select * from AA where id=:ID", use(2), into(p1);

**Comment by zann1x** [comment, marked as answer]
With `type_conversion`s, you're using SOCI's dynamic API. This API forces you to use the exact C++ data types that SOCI maps the database types to (see https://soci.sourceforge.net/doc/release/4.0/backends/odbc/#dynamic-binding).

While it is true that you're creating the column `id` as type `integer`, you have to keep in mind that you also created a `primary key` constraint on that column. Some databases treat such columns as something along the lines of `numeric(38,0)` internally, which of course would get mapped from `SQL_NUMERIC` to `double` by SOCI. I don't know how Oracle handles primary keys, but based on your description I would assume that this is exactly the issue here. Retrieving the column as `double` and then casting it to `int` might be your only option then.

**Reply by 330305020** [reply]
Thanks for your detailed explaination. It is exactly explained my question. Another column "class" as type integer have the same problem. Oracle treat integer as number(38, 0) as you said.
